Source/Core/Common/RangeSizeSet.h

```diff
@@ -0,0 +1,458 @@
+// Copyright 2020 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#pragma once
+
+#include <cassert>
+#include <map>
+
+namespace Common
+{
+// Like RangeSet, but additionally stores a map of the ranges sorted by their size, for quickly
+// finding the largest or smallest range.
+template <typename T>
+class RangeSizeSet
+{
+private:
+  struct Value;
+
+public:
+  // TODO: this should be an unsigned type when T is signed
```

Comment:
Just leaving this here so it's not forgotten about before merging, but this should be resolved.